plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java

```diff
@@ -292,7 +292,7 @@ public static Object convert(Object value, Type type)
             return ((ByteBuffer) value).array();
         }
         if (type instanceof Types.TimestampType) {
-            long epochMicros = (long) value;
+            long epochMicros = Long.parseLong(value.toString());
```

Comment:
When it's not `long`, what else is it?

Reply:
It could be Integer. see https://github.com/trinodb/trino/issues/7502, the partitioned column are converted to `DataType` in Iceberg, it will storage as `Integer`.

Reply:
Imo we are switching on the wrong type here then and should fix that instead. We are switching on the partition source column type here rather than the type of transform result (we that avilable to the callers so they can pass it). 

It's very unintuitive that a TimestampType branch (which Iceberg encodes as long) receives an int value because a transform is applied (day which Iceberg encodes as days since epoch as an int).

Reply:
Thanks for review, do you mean we should change the `convert` params to added one more type that indicate the actual storage type for each type? 
can you give me some more advices?

Reply:
https://github.com/trinodb/trino/pull/8101#issuecomment-852745163

Reply:
Thanks for the comments.
Here are some questions from my side: 
1. Is the semantics of the `convert` is that convert iceberg value to Trino value ?
2. If we pass the source type to the `convert`, then how to do the convert logic? In this case, we have already defined the type `Timestamp(6) with time zone`, that means we need to convert `DateType`(int) to `LongTimestampWithTimeZone`, but `DateType` should be convert to `DateType` intuitively.

Reply:
We shouldn't pass the source types to `PartitionTable#convert` - since it deals with partition columns it should only receive the values and types after the transform is applied.

If you fix the caller (`dataFileMatches`) then you just need to add a new branch here for handling `DateType`.

Reply:
Still have some doubts:
1.Does 'fix `dataFileMatches`' means pass the real type `DateType` into `PartitionTable#convert` ?
2.If so, as previous mentioned, we need to convert `DateType`(Integer) to `LongTimestampWithTimeZone`, how to convert? If we convert directly, then what can we do in case user really define the type `DateType` as partitioning column?

Reply:
>  1.Does 'fix dataFileMatches' means pass the real type DateType into PartitionTable#convert ?

We should the type of the transform result instead of type of source column.

> 2.If so, as previous mentioned, we need to convert DateType(Integer) to LongTimestampWithTimeZone, how to convert? If we convert directly, then what can we do in case user really define the type DateType as partitioning column?

You don't need to convert DateType to LongTimestampWithTimeZone - instead you add another case for `DateType` which will read the int and return as it is (because coincidentally Iceberg and Trino both represent dates in same manner - days since epoch. Correct me if wrong @findepi?).

Reply:
> > 2.If so, as previous mentioned, we need to convert DateType(Integer) to LongTimestampWithTimeZone, how to convert? If we convert directly, then what can we do in case user really define the type DateType as partitioning column?
> 
> You don't need to convert DateType to LongTimestampWithTimeZone - instead you add another case for `DateType` which will read the int and return as it is (because coincidentally Iceberg and Trino both represent dates in same manner - days since epoch. Correct me if wrong @findepi?).
If we return int it means, in https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsMaker.java#L221 , the `value` would be int, the `details.getTrinoType()` would be `DateType` in Trino, which still is long. we still need one more convert.

Reply:
@hashhar How about convert `DateType` into `Long` in `convert` method, just as you said we could add a `DateType` branch in the `convert`

Reply:
@hashhar @chenjian2664 we should definitely document (javadoc) what representation is expected as input & output of the `convert` method.
This whole conversation is precisely about that.

Reply:
BTW @chenjian2664 avoid conversions via `.toString()` or `Number` interface. They can mask correctness problems in the future.